overlord/snapstate/backend/mountunit_test.go

```diff
@@ -20,55 +20,55 @@
 package backend_test
 
 import (
+	"errors"
 	"fmt"
-	"os"
-	"path/filepath"
 
 	. "gopkg.in/check.v1"
 
 	"github.com/snapcore/snapd/dirs"
-	"github.com/snapcore/snapd/osutil"
-	"github.com/snapcore/snapd/osutil/squashfs"
 	"github.com/snapcore/snapd/overlord/snapstate/backend"
 	"github.com/snapcore/snapd/progress"
 	"github.com/snapcore/snapd/snap"
 	"github.com/snapcore/snapd/systemd"
+	"github.com/snapcore/snapd/systemd/systemdtest"
 	"github.com/snapcore/snapd/testutil"
 )
 
 type mountunitSuite struct {
-	umount *testutil.MockCmd
-
-	systemctlRestorer func()
-
 	testutil.BaseTest
 }
 
 var _ = Suite(&mountunitSuite{})
 
 func (s *mountunitSuite) SetUpTest(c *C) {
 	dirs.SetRootDir(c.MkDir())
-
-	// needed for system key generation
-	s.AddCleanup(osutil.MockMountInfo(""))
-
-	err := os.MkdirAll(filepath.Join(dirs.GlobalRootDir, "etc", "systemd", "system", "multi-user.target.wants"), 0755)
-	c.Assert(err, IsNil)
-
-	s.systemctlRestorer = systemd.MockSystemctl(func(cmd ...string) ([]byte, error) {
-		return []byte("ActiveState=inactive\n"), nil
-	})
-	s.umount = testutil.MockCommand(c, "umount", "")
 }
 
 func (s *mountunitSuite) TearDownTest(c *C) {
 	dirs.SetRootDir("")
-	s.umount.Restore()
-	s.systemctlRestorer()
 }
 
 func (s *mountunitSuite) TestAddMountUnit(c *C) {
-	restore := squashfs.MockNeedsFuse(false)
+	expectedErr := errors.New("creation error")
+
+	type mountUnitOptions struct {
+		name, revision, what, where, fstype string
+	}
+	var receivedParameters *mountUnitOptions
+	restore := backend.MockSystemdNew(func(mode systemd.InstanceMode, meter systemd.Reporter) systemd.Systemd {
+		sysd := &systemdtest.FakeSystemd{Mode: mode, Reporter: meter}
+		sysd.MockedAddMountUnitFile = func(name, revision, what, where, fstype string) (string, error) {
+			receivedParameters = &mountUnitOptions{
```

Comment:
I wonder if we should have a default implementation of this function even, something that just collects the parameters into something like `sysd.receivedParameters of type []MountUnitOptions` and maybe a `MockedAddMountUnitFileReturn` and `MockedAddMountUnitFileReturnErr`  so that there is less boilerplate to write (if its a common pattern).

Reply:
Such a default implementation would actually make sense for all the methods of this interface :-)

Maybe I should have a struct like
```
type MethodCall struct {
      name string
      params []interface{}
}
```
and then register all the invocations of all the methods? The only drawback I see with this approach, is that in the test code then we'd need to convert the `interface{}` types back into the real parameter type, in order to compare them...

Reply:
I'm not a fan of the params []interface approach.

Reply:
Another approach then could be this:
```
type ParamsForAddMountUnitFile {
    name, revision, what, where, fstype string
}
...
type ParamsFor<method name> {
   ....
}
```
and then,  the FakeSystem struct, we could store:
```
     addMountUnitFileInvocations []ParamsForAddMountUnitFile
     ...
     <method name>Invocations []ParamsFor<method name>

     // store also a list of all methods invoked, in the order they happen
     methodInvocations []string
```

So we would be able to check the parameters for all the invocations of a method, and also check that invocations of different methods have occurred in the desired order.